ML.NET to AI and Machine Learning (#43401)
dotnet/docs: docs/core/whats-new/dotnet-9/overview.md

Reviewer: Shouldn't we also add something about what's new in AI then too?

Reviewer 2: Working on making updates. to address #43311 

Will have those in a couple of hours.

For more information, see [What's new in the SDK for .NET 9](sdk.md).

> ## AI Building Blocks and Fundamentals

Reviewer: ```suggestion
## AI building blocks and fundamentals
```

### Microsoft.Extensions.AI & Microsoft.Extensions.VectorData

> .NET 9 introduces a unified layer of C# abstractions through [Microsoft.Extensions.AI](https://www.nuget.org/packages/Microsoft.Extensions.AI.Abstractions/) and [Microsoft.Extensions.VectorData](https://www.nuget.org/packages/Microsoft.Extensions.VectorData.Abstractions/). These abstractions facilitate interaction with AI services, including small and large language models (SLMs and LLMs), embeddings, vector stores, and middleware.

Reviewer: ```suggestion
.NET 9 introduces a unified layer of C# abstractions through the [Microsoft.Extensions.AI](https://www.nuget.org/packages/Microsoft.Extensions.AI.Abstractions/) and [Microsoft.Extensions.VectorData](https://www.nuget.org/packages/Microsoft.Extensions.VectorData.Abstractions/) packages. These abstractions facilitate interaction with AI services, including small and large language models (SLMs and LLMs), embeddings, vector stores, and middleware.
```

> ### Tokenizers

Reviewer: I have this under [What's new in ML.NET](https://learn.microsoft.com/en-us/dotnet/machine-learning/whats-new/overview). Should it not be under ML.NET?

Reviewer 3: oh nice... hmmmm yeah good point.... it all kinda blurs together to be honest.

> The [Microsoft.ML.Tokenizers](https://www.nuget.org/packages/Microsoft.ML.Tokenizers) library provides .NET developer with capabilities for encoding and decoding text to tokens. For AI scenarios, this is important to manage context, calculate cost, and pre-process text when working with local models.

Reviewer: ```suggestion
The [Microsoft.ML.Tokenizers](https://www.nuget.org/packages/Microsoft.ML.Tokenizers) library provides .NET developers with capabilities for encoding and decoding text to tokens. For AI scenarios, this is important to manage context, calculate cost, and preprocess text when working with local models.
```

The latest release introduces significant new capabilities:

- Tokenizers

>   - SentencePiece

Reviewer 4: ```suggest
    - Tiktoken for GPT (3, 3.5, 4, 4o, o1) and Llam3 models
    - Llama (based on SentencePiece) for Llama and Mistral models
    - CodeGen for code generation models like codegen-350M-mono
    - Phi2 (based on CodeGen) for Microsoft Phi2 model
    - WordPiece
    - Bert (based on WordPiece) for Bert supported models (e.g optimum--all-MiniLM-L6-v2)
```

#### TensorPrimitives

- **Expanded Method Scope:** Increased from 40 to nearly 200 overloads, now including numerical operations similar to `Math`, `MathF`, and `INumber<T>`, but for spans of values.
- **Performance Enhancements:** Many operations are now SIMD-optimized for better performance.
- **Generic Overloads:** Supports any T that implements a certain interface, expanding beyond just spans of float values in .NET.

#### Tensor

- Builds on top of `TensorPrimitives` for efficient math operations.
- Provides efficient interop with AI libraries (ML.NET, TorchSharp, ONNX Runtime) using zero copies where possible.
- Enables easy and efficient data manipulation with indexing and slicing operations.

## ML.NET

[ML.NET](https://www.nuget.org/packages/Microsoft.ML/) is an open-source, cross-platform framework that enables integration of custom machine-learning models into .NET applications.

ML.NET 4.0 brings the following improvements:

- New ways to programatically configure `MLContext` options.
- Load ONNX models as `Stream`.
- DataFrame improvements.
- (Experimental) TorchSharp ports of Llama and Phi family of models.

> - (Experimental) CausalLM pipeline APIs

Reviewer: This should link out to https://learn.microsoft.com/en-us/dotnet/machine-learning/whats-new/overview.

Reviewer 3: ```suggestion
- (Experimental) CausalLM pipeline APIs

For more information, see [What's new in ML.NET](dotnet/machine-learning/whats-new/overview).
```

> For more information, see [What's new in ML.NET](dotnet/machine-learning/whats-new/overview).

Reviewer: ```suggestion
For more information, see [What's new in ML.NET](../../../machine-learning/whats-new/overview.md).
```

For more information, see [What's new in the SDK for .NET 9](sdk.md).

## AI Building Blocks and Fundamentals

### Microsoft.Extensions.AI & Microsoft.Extensions.VectorData

.NET 9 introduces a unified layer of C# abstractions through [Microsoft.Extensions.AI](https://www.nuget.org/packages/Microsoft.Extensions.AI.Abstractions/) and [Microsoft.Extensions.VectorData](https://www.nuget.org/packages/Microsoft.Extensions.VectorData.Abstractions/). These abstractions facilitate interaction with AI services, including small and large language models (SLMs and LLMs), embeddings, vector stores, and middleware.

### Tokenizers

The [Microsoft.ML.Tokenizers](https://www.nuget.org/packages/Microsoft.ML.Tokenizers) library provides .NET developer with capabilities for encoding and decoding text to tokens. For AI scenarios, this is important to manage context, calculate cost, and pre-process text when working with local models.

The latest release introduces significant new capabilities:

- Tokenizers

>   - Byte-Level BPE

Reviewer 4: >Byte-Level BPE [](http://example.com/codeflow?start=4&length=14)

This one existed before so it is not introduced in the new release

  - SentencePiece

>   - WordPiece

Reviewer 4: Please list these as:

- Tiktoken
- Llama (based on SentencePiece)
- CodeGen
- Phi2 (based on CodeGen)
- WordPiece 
- Bert (based on WordPiece)

Reviewer 3: @tarekgh do you want to suggest the edits for this? I will accept them in what you have here looks good just want to make sure you get it looking lik eyou want

Reviewer 4: I added the suggestions. Sorry for the multiple as somehow the line selections didn't work when I added the first.

For more information, see [What's new in the SDK for .NET 9](sdk.md).

## AI Building Blocks and Fundamentals

### Microsoft.Extensions.AI & Microsoft.Extensions.VectorData

.NET 9 introduces a unified layer of C# abstractions through [Microsoft.Extensions.AI](https://www.nuget.org/packages/Microsoft.Extensions.AI.Abstractions/) and [Microsoft.Extensions.VectorData](https://www.nuget.org/packages/Microsoft.Extensions.VectorData.Abstractions/). These abstractions facilitate interaction with AI services, including small and large language models (SLMs and LLMs), embeddings, vector stores, and middleware.

### Tokenizers

The [Microsoft.ML.Tokenizers](https://www.nuget.org/packages/Microsoft.ML.Tokenizers) library provides .NET developer with capabilities for encoding and decoding text to tokens. For AI scenarios, this is important to manage context, calculate cost, and pre-process text when working with local models.

The latest release introduces significant new capabilities:

- Tokenizers

>   - Byte-Level BPE

Reviewer 3: ```suggestion
```

  - SentencePiece
  - WordPiece
- Built-in tokenizers for the following models:
  - GPT (3, 3.5, 4, 4o, o1)
  - Llama
  - Phi
  - BERT
  - CodeGen

### Tensors

In .NET 9, `TensorPrimitives` and the new `Tensor<T>` type expand AI capabilities by enabling efficient encoding, manipulation, and computation of multi-dimensional data.

Improvements in the latest release of [System.Numerics.Tensors](https://www.nuget.org/packages/System.Numerics.Tensors/) include:

#### TensorPrimitives

- **Expanded Method Scope:** Increased from 40 to nearly 200 overloads, now including numerical operations similar to `Math`, `MathF`, and `INumber<T>`, but for spans of values.
- **Performance Enhancements:** Many operations are now SIMD-optimized for better performance.

> - **Generic Overloads:** Supports any T that implements a certain interface, expanding beyond just spans of float values in .NET 

Reviewer 3: ```suggestion
- **Generic Overloads:** Supports any T that implements a certain interface, expanding beyond just spans of float values in .NET.
```

For more information, see [What's new in the SDK for .NET 9](sdk.md).

## AI building blocks and fundamentals

### Microsoft.Extensions.AI & Microsoft.Extensions.VectorData

.NET 9 introduces a unified layer of C# abstractions through the [Microsoft.Extensions.AI](https://www.nuget.org/packages/Microsoft.Extensions.AI.Abstractions/) and [Microsoft.Extensions.VectorData](https://www.nuget.org/packages/Microsoft.Extensions.VectorData.Abstractions/) packages. These abstractions facilitate interaction with AI services, including small and large language models (SLMs and LLMs), embeddings, vector stores, and middleware.

### Tokenizers

The [Microsoft.ML.Tokenizers](https://www.nuget.org/packages/Microsoft.ML.Tokenizers) library provides .NET developers with capabilities for encoding and decoding text to tokens. For AI scenarios, this is important to manage context, calculate cost, and preprocess text when working with local models.

The latest release introduces significant new capabilities:

- Tokenizers
  - SentencePiece

>   - WordPiece

Reviewer 4: ```suggestion
```

> - Built-in tokenizers for the following models:
>   - GPT (3, 3.5, 4, 4o, o1)
>   - Llama
>   - Phi
>   - BERT
>   - CodeGen

Reviewer 4: ```suggestion
```
